dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,837 @@
+/*
+ * PROJECT:     ReactOS CabView Shell Extension
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Shell folder implementation
+ * COPYRIGHT:   Copyright 2024 Whindmar Saksit <whindsaks@proton.me>
+ */
+
+#include "cabview.h"
+#include "util.h"
+
+enum FOLDERCOLUMNS
+{
+    COL_NAME,   // PKEY_ItemNameDisplay
+    COL_SIZE,   // PKEY_Size
+    COL_TYPE,   // PKEY_ItemTypeText
+    COL_MDATE,  // PKEY_DateModified
+    COL_PATH,   // PKEY_?: Archive-relative path
+    COL_ATT,    // PKEY_FileAttributes
+    COLCOUNT
+};
+
+static const struct FOLDERCOLUMN
+{
+    BYTE TextId;
+    BYTE LvcFmt;
+    BYTE LvcChars;
+    BYTE ColFlags;
+    const GUID *pkg;
+    BYTE pki;
+} g_Columns[] =
+{
+    { IDS_COL_NAME, LVCFMT_LEFT, 20, SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT, &FMTID_Storage, PID_STG_NAME },
+    { IDS_COL_SIZE, LVCFMT_RIGHT, 16, SHCOLSTATE_TYPE_INT | SHCOLSTATE_ONBYDEFAULT, &FMTID_Storage, PID_STG_SIZE },
+    { IDS_COL_TYPE, LVCFMT_LEFT, 20, SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT, &FMTID_Storage, PID_STG_STORAGETYPE },
+    { IDS_COL_MDATE, LVCFMT_LEFT, 20, SHCOLSTATE_TYPE_DATE | SHCOLSTATE_ONBYDEFAULT, &FMTID_Storage, PID_STG_WRITETIME },
+    { IDS_COL_PATH, LVCFMT_LEFT, 30, SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT, &CLSID_CabFolder, 0 },
+    { IDS_COL_ATT, LVCFMT_RIGHT, 10, SHCOLSTATE_TYPE_STR, &FMTID_Storage, PID_STG_ATTRIBUTES },
+};
+
+static CABITEM* CreateItem(LPCWSTR Path, UINT Attrib, UINT Size, UINT DateTime)
+{
+    const SIZE_T len = lstrlenW(Path), cb = FIELD_OFFSET(CABITEM, Path[len + 1]);
+    if (cb > 0xffff)
+        return NULL;
+    CABITEM *p = (CABITEM*)SHAlloc(cb + sizeof(USHORT));
+    if (p)
+    {
+        p->cb = (USHORT)cb;
+        p->Unknown = 0;
+        p->Size = Size;
+        p->Attrib = Attrib;
+        p->Date = HIWORD(DateTime);
+        p->Time = LOWORD(DateTime);
+        p->NameOffset = 0;
+        for (UINT i = 0;; ++i)
+        {
+            WCHAR c = Path[i];
+            if (c == L':') // Don't allow absolute paths
+                c = L'_';
+            if (c == L'/') // Normalize
+                c = L'\\';
```

Comment:
I have seen this somewhere before in this PR, would it make sense to share the code, or would that make it less readable?